dashboard-metrics-report-generator/src/main/scala/it/pagopa/interop/dashboardmetricsreportgenerator/Jobs.scala

```diff
@@ -105,21 +109,30 @@ object Jobs {
   def getTokensData(readModel: FileManager, config: TokensBucketConfiguration): Future[TokensData] = {
 
     def getIssueDate(token: String): Future[OffsetDateTime] =
-      Future(token.parseJson)
-        .flatMap(js => Future(js.asJsObject))
+      Future(token.parseJson.asJsObject)
         .flatMap(_.fields.get("issuedAt").toFuture(GenericError("Missing issuedAt field in token")))
         .flatMap {
-          case JsNumber(value) => Future.successful(value.toLong)
+          case JsNumber(value) => value.toLong.toOffsetDateTime.toFuture
           case _               => Future.failed(GenericError("issuedAt should be a number"))
         }
-        .flatMap(_.toOffsetDateTime.toFuture)
 
-    readModel
-      .getAllFiles(config.bucket)(config.basePath)
-      .map(_.values.flatMap(new String(_).split('\n')).toList)
-      .flatMap(tokens => Future.traverse(tokens)(getIssueDate))
-      .map { issueTimes =>
-        val twoWeeksAgo: OffsetDateTime = OffsetDateTime.now().minusDays(14)
+    def allTokensPaths(): Future[List[readModel.StorageFilePath]] =
+      readModel.listFiles(config.bucket)(config.basePath)
+
+    def getTokens(path: String): Future[List[String]] =
+      readModel.getFile(config.bucket)(path).map(bs => new String(bs).split('\n').toList)
+
+    val twoWeeksAgo: OffsetDateTime = OffsetDateTime.now().minusDays(14)
+
+    allTokensPaths()
+      .flatMap(paths =>
+        Future.traverseWithLatch(maxParallelism)(paths)(path =>
+          // * This is safe to do w/o max parallelism as it's not blocking
```

Comment:
not sure I got the reason behind this, but I trust your tests

Reply:
The main reason is the fact that once limited to half a CPU the first error we got was an aws timeout for file getting on s3. Now I'm using `traverseWithLatch` and `maxParallelism` set to `Runtime.getRuntime().availableProcessors() - 1` in order to be sure that the max number of GETed files are at a time are $maxParallelism and so the available CPUs are free to block while waiting the file contents w/o having a timeout running for all the other unscheduled calls.